qiskit/assertions/assertproduct.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Assertion of product states.
+"""
+import numpy as np
+from scipy.stats import chi2_contingency
+from qiskit.assertions.asserts import Asserts
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.quantumcircuit import Register
+
+
+class AssertProduct(Asserts):
+    """
+        A measurement instruction that additionally performs statistical tests on the measurement
+        outcomes to assert whether the state is a product state or not.
+    """
+    def __init__(self, qubit0, cbit0, qubit1, cbit1, pcrit, negate):
+        """
+        Constructor for AssertProduct
+
+        Args:
+            qubit0(QuantumRegister or list or tuple): quantum register
+            cbit0(ClassicalRegister or list or tuple): classical register
+            qubit1(QuantumRegister or list or tuple): quantum register
+            cbit1(ClassicalRegister or list or tuple): classical register
+            pcrit(float): the critical p-value
+            negate(bool): True if assertion passed is negation of statistical test passed
+        """
+        self._qubit0 = self.syntax4measure(qubit0)
+        self._cbit0 = self.syntax4measure(cbit0)
+        self._qubit1 = self.syntax4measure(qubit1)
+        self._cbit1 = self.syntax4measure(cbit1)
+        type_str = "Not Product" if negate else "Product"
```

Comment:
Maybe have classes for different assertion types?

Reply:
There are different classes; each assertion type (AssertClassical, AssertProduct, and AssertUniform) and its logical negation are the same type except for the value of the boolean negate.
The type string is for the user to print out what type of assertion they had run, so given an assertion result (passed or failed), they can easily tell whether it was a "Not Product" assertion or a "Product" assertion.